lib/kafka/producer.rb

```diff
@@ -357,6 +365,7 @@ def assign_partitions!
             exception: [e.class.to_s, e.message],
           })
 
+          topics_with_failures << message.topic
           failed_messages << message
```

Comment:
Other messages for the same topic where a message failed to receive a partition are kept in `failing_messages` which is ultimately copied back to `@pending_message_queue`.

In the loop in `#deliver_messages_with_retries`, the `@pending_message_queue` size is included in `#buffer_size`. Since the size is still non-zero the loop is retried (if still below the max retries configuration).

So the partition assignment will be retried for the originally failing message and all subsequent messages for the topic.